<a href="https://colab.research.google.com/github/OlgaZy/Python_for_DA/blob/main/HW_9_2_%D0%9E%D1%81%D0%BD%D0%BE%D0%B2%D0%B8_%D1%80%D0%BE%D0%B1%D0%BE%D1%82%D0%B8_%D0%B7_%D0%B4%D0%B0%D0%BD%D0%B8%D0%BC%D0%B8_%D1%83_Pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Домашнє завдання до теми "Основи роботи з даними у Pandas"

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [50]:
import pandas as pd

In [51]:
df = pd.read_csv('/content/drive/MyDrive/DA_in_IT/Data/supermarket_sales.csv')

In [49]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


1. За допомогою бібліотеки Pandas виведіть перші 4 і останні 4 записи у наборі даних (це дві окремі команди). Треба саме 4! Не 5.

In [52]:
df.head(4)

Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   
2  631-41-3108      A     Yangon        Normal    Male   
3  123-19-1176      A     Yangon        Member    Male   

             Product line  Unit price  Quantity   Tax 5%     Total       Date  \
0       Health and beauty       74.69         7  26.1415  548.9715   1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200   3/8/2019   
2      Home and lifestyle       46.33         7  16.2155  340.5255   3/3/2019   
3       Health and beauty       58.22         8  23.2880  489.0480  1/27/2019   

    Time      Payment    cogs  gross margin percentage  gross income  Rating  
0  13:08      Ewallet  522.83                 4.761905       26.1415     9.1  
1  10:29         Cash   76.40                 4.761905        3.8200     9.6  
2  13:23  Credit card  324.31                 4.761905       16.2155     7.4  
3  20:33      Ewallet  465.76                 4.761905       23.2880     8.4

In [53]:
df.tail(4)

Invoice ID Branch      City Customer type  Gender         Product line  \
996  303-96-2227      B  Mandalay        Normal  Female   Home and lifestyle   
997  727-02-1313      A    Yangon        Member    Male   Food and beverages   
998  347-56-2442      A    Yangon        Normal    Male   Home and lifestyle   
999  849-09-3807      A    Yangon        Member  Female  Fashion accessories   

     Unit price  Quantity  Tax 5%     Total       Date   Time  Payment  \
996       97.38        10  48.690  1022.490   3/2/2019  17:16  Ewallet   
997       31.84         1   1.592    33.432   2/9/2019  13:22     Cash   
998       65.82         1   3.291    69.111  2/22/2019  15:33     Cash   
999       88.34         7  30.919   649.299  2/18/2019  13:28     Cash   

       cogs  gross margin percentage  gross income  Rating  
996  973.80                 4.761905        48.690     4.4  
997   31.84                 4.761905         1.592     7.7  
998   65.82                 4.761905         3.291     4.1  
999  618.38                 4.761905        30.919     6.6

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1F5XlzBOOTMsedpsYqz6NnXjIPTDe6jpx)

2. Виведіть рядки даних з індексами з 615 по 619 включно.

In [54]:
df[615:620]

Invoice ID Branch       City Customer type  Gender  \
615  509-29-3912      A     Yangon        Member  Female   
616  327-40-9673      B   Mandalay        Member    Male   
617  840-19-2096      C  Naypyitaw        Member    Male   
618  828-46-6863      A     Yangon        Member    Male   
619  641-96-3695      C  Naypyitaw        Member  Female   

               Product line  Unit price  Quantity   Tax 5%     Total  \
615       Sports and travel       38.72         9  17.4240  365.9040   
616       Sports and travel       72.60         6  21.7800  457.3800   
617  Electronic accessories       87.91         5  21.9775  461.5275   
618      Food and beverages       98.53         6  29.5590  620.7390   
619     Fashion accessories       43.46         6  13.0380  273.7980   

          Date   Time      Payment    cogs  gross margin percentage  \
615  3/20/2019  12:24      Ewallet  348.48                 4.761905   
616  1/13/2019  19:51         Cash  435.60                 4.761905   
617  3/14/2019  18:10      Ewallet  439.55                 4.761905   
618  1/23/2019  11:22  Credit card  591.18                 4.761905   
619   2/7/2019  17:55      Ewallet  260.76                 4.761905   

     gross income  Rating  
615       17.4240     4.2  
616       21.7800     6.9  
617       21.9775     4.4  
618       29.5590     4.0  
619       13.0380     8.5

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1Otufe_TXRmjZxWigzBiVBLCAVUfEReXM)

3. Виведіть назви колонок у наборі даних разом з їхніми типами. Це можна зробити кількома способами, важливо аби в результаті були назви колонок і відповідні їм типи.

In [55]:
df.dtypes

Invoice ID                  object
Branch                      object
City                        object
Customer type               object
Gender                      object
Product line                object
Unit price                 float64
Quantity                     int64
Tax 5%                     float64
Total                      float64
Date                        object
Time                        object
Payment                     object
cogs                       float64
gross margin percentage    float64
gross income               float64
Rating                     float64
dtype: object

In [56]:
type(df.iloc[0].Payment)

str

Очікуваний результат (один з можливих форматів):

![](https://drive.google.com/uc?export=view&id=1ASh6-S-_f3X9fTQAPIvnHMFZamTBjNkL)

4. Користуючись методом `DataFrame.describe()` виявіть 1, 25, 50, 75 і 99 персентилі колонки `gross income`. Знайдіть, яке значення відповідає 75отму персентилю цієї колонки і що це означає? Як би Ви його пояснили простими словами колегам? Відповідь напишіть текстом в окремій клітинці типу Markdown.

In [57]:
percentiles = df['gross income'].describe(percentiles=[.01, .25, .5, .75, .99])
percentiles

count    1000.000000
mean       15.379369
std        11.708825
min         0.508500
1%          0.913735
25%         5.924875
50%        12.088000
75%        22.445250
99%        45.250750
max        49.650000
Name: gross income, dtype: float64

75й персентиль означає, що 75% даних, в нашому випадку це показник валового доходу з інвойсу, є меншими або дорівнюють 22.45. Значить, що решта 25% відсотків чеків приносять високий дохід.

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1bDa1YFUWDsPm78Lnpy2geQND8M1HEQiT)

5. Виведіть кількість інвойсів за кожним унікальним містом в наборі даних.

  Підказка: інвойс - це один рядок в цьому наборі даних, ми можемо цим скористатись аби швидко отримати відповідь.

In [58]:
df.groupby('City')['Invoice ID'].count()

City
Mandalay     332
Naypyitaw    328
Yangon       340
Name: Invoice ID, dtype: int64

In [59]:
df.City.value_counts()

City
Yangon       340
Mandalay     332
Naypyitaw    328
Name: count, dtype: int64

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1G8c6M875NOW7oU3OtZegTOnyD2y8lhAV)

6. Виведіть, який відсоток інвойсів припадає на кожне унікальне місто.

In [61]:
pers_of_invoces = df.City.value_counts() / df.City.value_counts().sum() * 100
pers_of_invoces

City
Yangon       34.0
Mandalay     33.2
Naypyitaw    32.8
Name: count, dtype: float64

In [62]:
df.City.value_counts(normalize=True)

City
Yangon       0.340
Mandalay     0.332
Naypyitaw    0.328
Name: proportion, dtype: float64

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1yZZpWhL2y8DX9rSRioYCRh1VbJgcf3fY)

7.1. Використовуючи фільтрацію даних, знайдіть середнє значення `gross income` для міст `Yangon` та `Mandalay`. В якому місті воно більше?

7.2 Сформулюйте 3 гіпотези, чому в одному місті gross income вище, як можна дізнатись чи правдива ця гіпотеза з використанням даних (наявних і додаткових), та як висновки з цих гіпотез можна використати для прийняття рішень як підвищити прибуток в мережі супермаркетів. Відповідь напишіть текстом в окремій клітинці типу Markdown.

Це завдання - на продуктове мислення. Якщо не знаєте, які гіпотези сформулювати - напишіть ті, які здаються вам зовсім примітивними. Важливо навчитись висловлювати свої гіпотези щодо даних, і з часом та досвідом вони будуть ставати щоразу глибше і точніше. Весь аналіз даних будується на тому, що ми перевіряємо певні гіпотези та робимо висновки і рекоемндації на основі своїх досліджень.

In [63]:
df.groupby('City')['gross income'].mean()

City
Mandalay     15.232024
Naypyitaw    16.052367
Yangon       14.874001
Name: gross income, dtype: float64

In [19]:
filter = (df.City == 'Mandalay') | (df.City == 'Yangon')
df[filter].groupby('City')['gross income'].mean()

City
Mandalay    15.232024
Yangon      14.874001
Name: gross income, dtype: float64

Бачимо, що в місті Mandalay середнє значення gross income більше, та становить 15.23.

Гіпотези:
1. У місті Mandalay купують дорожчі товари або більшу кількість товарів. Треба порівняти середню ціну товару та середню кількість товарів у чеку.
2. Краще працює система лояльності і більше постійних покупців. Треба порахувати відсоток покупців з типом Customer type = Member по кожному місту. Порівняти середній gross income постійних клієнтів.
3. Наявність викидів. Перевірити чи є чеки на дуже великі суми, які можуть впливати на рівень середньго.
4. Порівняти націнку на товари в різних містах.





Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1JdXUmK46WcoXUJYP7NS2SJlr2XjBbpQ8)

8. Для продовження дослідження різниці у прибутках в магазинах давайте зʼясуємо, чи відрізняється маржа ('gross margin percentage') для різних інвойсів в магазині? Перевірте це програмно на даних виходячи з того функціоналу, який ми вже пройшли у лекціях.

Дайте на основі отриманого результату відповідь: "У всіх всіх інвойсах однакова маржа" або "Маржа відрізняється для різних інвойсів".

In [64]:
df.groupby('City')['gross margin percentage'].unique()

City
Mandalay     [4.761904762]
Naypyitaw    [4.761904762]
Yangon       [4.761904762]
Name: gross margin percentage, dtype: object

In [65]:
df['gross margin percentage'].unique()

array([4.76190476])

У всіх всіх інвойсах однакова маржа

9. Врахуовуючи висновок в попередньому завданні, дослідимо як продаються в цих містах різні категорії товарів. Виведіть відсотковий розпоіділ кількості інвойсів по категоріям товарів `Product line` в містах `Yangon` та `Mandalay`.

In [66]:
filter2 = (df.City == 'Mandalay') | (df.City == 'Yangon')
grouped_invoces = df[filter2].groupby(['Product line', 'City'])['Invoice ID'].count()
grouped_invoces

Product line            City    
Electronic accessories  Mandalay    55
                        Yangon      60
Fashion accessories     Mandalay    62
                        Yangon      51
Food and beverages      Mandalay    50
                        Yangon      58
Health and beauty       Mandalay    53
                        Yangon      47
Home and lifestyle      Mandalay    50
                        Yangon      65
Sports and travel       Mandalay    62
                        Yangon      59
Name: Invoice ID, dtype: int64

In [67]:
total_invoces = df['Invoice ID'].count()
percent_of_total_invoces = grouped_invoces / total_invoces * 100
percent_of_total_invoces

Product line            City    
Electronic accessories  Mandalay    5.5
                        Yangon      6.0
Fashion accessories     Mandalay    6.2
                        Yangon      5.1
Food and beverages      Mandalay    5.0
                        Yangon      5.8
Health and beauty       Mandalay    5.3
                        Yangon      4.7
Home and lifestyle      Mandalay    5.0
                        Yangon      6.5
Sports and travel       Mandalay    6.2
                        Yangon      5.9
Name: Invoice ID, dtype: float64

In [68]:
total_city_invoces = df[filter2].groupby(['City'])['Invoice ID'].count()
total_city_invoces

City
Mandalay    332
Yangon      340
Name: Invoice ID, dtype: int64

In [69]:
  percent_of_invoces_Mandalay = df[(df.City == 'Mandalay')].groupby(['Product line'])['Invoice ID'].count() / df[(df.City == 'Mandalay')]['Invoice ID'].count() * 100
  percent_of_invoces_Mandalay

Product line
Electronic accessories    16.566265
Fashion accessories       18.674699
Food and beverages        15.060241
Health and beauty         15.963855
Home and lifestyle        15.060241
Sports and travel         18.674699
Name: Invoice ID, dtype: float64

In [70]:
  percent_of_invoces_Yangon = df[(df.City == 'Yangon')].groupby(['Product line'])['Invoice ID'].count() / df[(df.City == 'Yangon')]['Invoice ID'].count() * 100
  percent_of_invoces_Yangon

Product line
Electronic accessories    17.647059
Fashion accessories       15.000000
Food and beverages        17.058824
Health and beauty         13.823529
Home and lifestyle        19.117647
Sports and travel         17.352941
Name: Invoice ID, dtype: float64

Очікуваний результат для кожного з міст:

![](https://drive.google.com/uc?export=view&id=1IbG58YK5KVqC-wCnllF1btgEgfWTbi_O)

![](https://drive.google.com/uc?export=view&id=1aKOkLmlMq1sOh4GbwyaQyjfhNJzdU4Yh)